viewing_party/party.py

```diff
@@ -1,23 +1,113 @@
+    average = 0
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    for movie in user_data["watched"]: 
+        sum += movie["rating"]
+    average = sum / len(user_data["watched"])
+    return average
+
+def create_frequency_counter(user_data):
+    frequency_counter = {}
+    for movie in user_data["watched"]:
+        if movie["genre"] not in frequency_counter:
+            frequency_counter[movie["genre"]] = 1
+        else:
+            frequency_counter[movie["genre"]] += 1
+    return frequency_counter
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+            return None
+    frequency_counter = create_frequency_counter(user_data)
+    mode = 0
+    mode_genre = ""
+    for genre_name, genre_frequency in frequency_counter.items():
+        if genre_frequency > mode:
+            mode = genre_frequency
+            mode_genre = genre_name
+    return mode_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_friends_unique_watched(user_data):
+    result = []
+    for friends in user_data["friends"]:
+        for friends_watched_movies in friends["watched"]:
+            if friends_watched_movies not in user_data["watched"] and friends_watched_movies not in result:
+                result.append(friends_watched_movies)
+    return result
+
+def get_unique_watched(user_data):
+    result = []
+    friends_movies_watch = []
+    for friend in user_data["friends"]:
+        for friends_watched_movie in friend["watched"]:
+            friends_movies_watch.append(friends_watched_movie)
+    for movie in user_data["watched"]:
+        if movie not in friends_movies_watch:
+            result.append(movie)
+    return result
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommendations = []
+    movies_to_watch = get_friends_unique_watched(user_data)
+    for movies in movies_to_watch:
+        if movies["host"] in user_data["subscriptions"]:
+            recommendations.append(movies)
+    return recommendations
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    movie_recs = []
+    most_watched_genre = get_most_watched_genre(user_data)
+    unique_watched_movies = get_friends_unique_watched(user_data)
+    for movie in unique_watched_movies:
+        if movie["genre"] == most_watched_genre:
+            movie_recs.append(movie)
+    return movie_recs
+
+def get_rec_from_favorites(user_data):
+    movie_recs = []
+    user_unique_watched_movies = get_unique_watched(user_data)
+    for movie in user_data["favorites"]:
+        if movie in user_unique_watched_movies:
+            movie_recs.append(movie)
+    return movie_recs

```

Comment:
👍  Nice work! We can also use list comprehension here to build `movie_recs`.

Reply:
always forgetting the ol list comprehension these days!